File: solutions/java/maximum-depth-of-binary-tree.java
// id: 011
// title: Maximum Depth of Binary Tree
// approach: DFS Recursion (level-order input)
// time: O(n) | space: O(n)
// author: chron303

import java.io.*;

public class Main {
    static int[] tree;

    static int maxDepth(int i) {
        if (i >= tree.length || tree[i] == -1) return 0;
        return 1 + Math.max(maxDepth(2 * i + 1), maxDepth(2 * i + 2));
    }

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine().trim());

        if (n == 0) {
            System.out.println(0);
            return;
        }

        String[] input = br.readLine().trim().split(" ");
        tree = new int[n];
        for (int i = 0; i < n; i++) {
            tree[i] = Integer.parseInt(input[i]);
        }

        System.out.println(maxDepth(0));
    }
}

File: solutions/python/maximum-depth-of-binary-tree.py
# id: 011
# title: Maximum Depth of Binary Tree
# approach: DFS Recursion (level-order input)
# time: O(n) | space: O(n)
# author: chron303

import sys
sys.setrecursionlimit(10000)

def max_depth(tree, i):
    if i >= len(tree) or tree[i] == -1:
        return 0
    return 1 + max(max_depth(tree, 2 * i + 1), max_depth(tree, 2 * i + 2))

def main():
    data = sys.stdin.read().split()
    n = int(data[0])

    if n == 0:
        print(0)
        return

    tree = list(map(int, data[1:n+1]))
    print(max_depth(tree, 0))

if __name__ == "__main__":
    main()
